File: crates/aft/src/commands/semantic_eval.rs
//! `semantic_eval` command — run a local JSONL eval suite against AFT's
//! semantic search and report recall@k and MRR.
//!
//! ## Wire format
//!
//! Request:
//! ```json
//! {
//!   "path": ".aft/semantic-eval.jsonl",
//!   "top_k": 10,
//!   "include_per_case": true
//! }
//! ```
//!
//! - `path` (required) — JSONL file. Each line is one eval case.
//! - `top_k` (optional) — default cutoff for recall@k (default 10).
//! - `include_per_case` (optional, default true) — include per-case results
//!   in the response. Set false for a one-line summary in agent output.
//!
//! ## Response
//!
//! ```json
//! {
//!   "total": 12,
//!   "hits_in_top_k": 9,
//!   "recall_at_k": 0.75,
//!   "mrr": 0.612,
//!   "k": 10,
//!   "cases": [ { "index": 0, "query": "...", "first_hit_rank": 1, ... } ]
//! }
//! ```
//!
//! Or when `include_per_case` is false:
//! ```json
//! { "summary_line": "eval: 9/12 hits, recall@10=0.750, mrr=0.612" }
//! ```

use serde::Deserialize;

use crate::protocol::{RawRequest, Response};
use crate::semantic_eval as eval;

#[derive(Debug, Deserialize)]
struct SemanticEvalParams {
    path: String,
    #[serde(default = "default_top_k")]
    top_k: usize,
    #[serde(default = "default_include_per_case")]
    include_per_case: bool,
}

fn default_top_k() -> usize {
    10
}
fn default_include_per_case() -> bool {
    true
}

pub fn handle_semantic_eval(req: &RawRequest, _ctx: &crate::context::AppContext) -> Response {
    let params: SemanticEvalParams = match serde_json::from_value(req.params.clone()) {
        Ok(p) => p,
        Err(e) => {
            return Response::error(
                &req.id,
                "invalid_request",
                format!("semantic_eval: invalid params: {e}"),
            );
        }
    };
    if params.top_k == 0 {
        return Response::error(
            &req.id,
            "invalid_request",
            "semantic_eval: top_k must be >= 1".to_string(),
        );
    }
    let text = match std::fs::read_to_string(&params.path) {
        Ok(t) => t,
        Err(e) => {
            return Response::error(
                &req.id,
                "eval_file_unreadable",
                format!("semantic_eval: cannot read {}: {e}", params.path),
            );
        }
    };
    let cases = match eval::parse_jsonl(&text) {
        Ok(c) => c,
        Err(e) => {
            return Response::error(
                &req.id,
                "eval_file_parse_error",
                format!("semantic_eval: {e}"),
            );
        }
    };
    // Note: This stub returns zero retrieved hits per case. Wiring to
    // `handle_semantic_search` is deferred to a follow-up Bead; for now the
    // harness is exercised through its pure-logic surface (parser, matcher,
    // scorer). Misses surface as expected and are the agent's signal that
    // the upstream wiring is not yet in place.
    let results: Vec<Vec<eval::RetrievedHit>> = cases.iter().map(|_| Vec::new()).collect();
    let summary = eval::score_suite(&cases, &results, params.top_k);

    let mut payload = serde_json::json!({
        "total": summary.total,
        "hits_in_top_k": summary.hits_in_top_k,
        "recall_at_k": summary.recall_at_k,
        "mrr": summary.mrr,
        "k": summary.k,
        "summary_line": summary.render_line(),
    });
    if params.include_per_case {
        payload["cases"] =
            serde_json::to_value(&summary.cases).unwrap_or(serde_json::Value::Array(vec![]));
    }
    Response::success(&req.id, payload)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::Config;
    use crate::context::AppContext;
    use crate::parser::TreeSitterProvider;
    use crate::protocol::RawRequest;
    use serde_json::json;

    fn req_for(params: serde_json::Value) -> RawRequest {
        RawRequest {
            id: "test-1".to_string(),
            command: "semantic_eval".to_string(),
            lsp_hints: None,
            session_id: None,
            params,
        }
    }

    fn make_ctx() -> AppContext {
        AppContext::new(Box::new(TreeSitterProvider::new()), Config::default())
    }

    use std::sync::atomic::{AtomicU64, Ordering};

    static EVAL_FILE_COUNTER: AtomicU64 = AtomicU64::new(0);

    fn write_eval(content: &str) -> std::path::PathBuf {
        let counter = EVAL_FILE_COUNTER.fetch_add(1, Ordering::Relaxed);
        let dir =
            std::env::temp_dir().join(format!("aft-eval-test-{}-{}", std::process::id(), counter));
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("eval.jsonl");
        std::fs::write(&path, content).unwrap();
        path
    }

    #[test]
    fn handle_returns_summary_for_valid_eval() {
        let path = write_eval(
            r#"{"query":"q1","expected_paths":["a.rs"]}
{"query":"q2","expected_paths":["b.rs"]}
"#,
        );
        let req = req_for(json!({ "path": path.to_string_lossy() }));
        let ctx = make_ctx();
        let resp = handle_semantic_eval(&req, &ctx);
        assert!(resp.success, "got: {resp:?}");
        let v = &resp.data;
        assert_eq!(v["total"], 2);
        assert_eq!(v["hits_in_top_k"], 0); // stub returns no hits
        assert_eq!(v["k"], 10);
        assert!(v["summary_line"].as_str().unwrap().contains("0/2"));
    }

    #[test]
    fn handle_rejects_missing_path_param() {
        let req = req_for(json!({}));
        let ctx = make_ctx();
        let resp = handle_semantic_eval(&req, &ctx);
        assert!(!resp.success);
        assert_eq!(resp.data["code"], "invalid_request");
    }

    #[test]
    fn handle_rejects_unreadable_path() {
        let req = req_for(json!({ "path": "/nonexistent/path/to/eval.jsonl" }));
        let ctx = make_ctx();
        let resp = handle_semantic_eval(&req, &ctx);
        assert!(!resp.success);
        assert_eq!(resp.data["code"], "eval_file_unreadable");
    }

    #[test]
    fn handle_rejects_zero_top_k() {
        let path = write_eval(r#"{"query":"q1","expected_paths":["a.rs"]}"#);
        let req = req_for(json!({ "path": path.to_string_lossy(), "top_k": 0 }));
        let ctx = make_ctx();
        let resp = handle_semantic_eval(&req, &ctx);
        assert!(!resp.success);
        assert_eq!(resp.data["code"], "invalid_request");
    }

    #[test]
    fn handle_rejects_invalid_jsonl() {
        let path = write_eval("not json\n");
        let req = req_for(json!({ "path": path.to_string_lossy() }));
        let ctx = make_ctx();
        let resp = handle_semantic_eval(&req, &ctx);
        assert!(!resp.success);
        assert_eq!(resp.data["code"], "eval_file_parse_error");
    }

    #[test]
    fn handle_omits_per_case_when_disabled() {
        let path = write_eval(r#"{"query":"q1","expected_paths":["a.rs"]}"#);
        let req = req_for(json!({
            "path": path.to_string_lossy(),
            "include_per_case": false
        }));
        let ctx = make_ctx();
        let resp = handle_semantic_eval(&req, &ctx);
        let v = &resp.data;
        assert!(v.get("cases").is_none(), "got: {v}");
        assert!(v.get("summary_line").is_some());
    }

    #[test]
    fn handle_includes_per_case_by_default() {
        let path = write_eval(r#"{"query":"q1","expected_paths":["a.rs"]}"#);
        let req = req_for(json!({ "path": path.to_string_lossy() }));
        let ctx = make_ctx();
        let resp = handle_semantic_eval(&req, &ctx);
        let v = &resp.data;
        assert!(v.get("cases").is_some(), "got: {v}");
        let cases = v["cases"].as_array().unwrap();
        assert_eq!(cases.len(), 1);
        assert_eq!(cases[0]["query"], "q1");
    }

    #[test]
    fn handle_respects_top_k_override() {
        let path = write_eval(r#"{"query":"q1","expected_paths":["a.rs"]}"#);
        let req = req_for(json!({
            "path": path.to_string_lossy(),
            "top_k": 3
        }));
        let ctx = make_ctx();
        let resp = handle_semantic_eval(&req, &ctx);
        let v = &resp.data;
        assert_eq!(v["k"], 3);
    }
}

File: packages/opencode-plugin/src/tools/semantic-eval.ts
import type { ToolDefinition } from "@opencode-ai/plugin";
import { tool } from "@opencode-ai/plugin";
import type { PluginContext } from "../types.js";
import { callBridge } from "./_shared.js";

const z = tool.schema;

type ToolArg = ToolDefinition["args"][string];

function arg(schema: unknown): ToolArg {
  return schema as ToolArg;
}

/**
 * aft_semantic_eval — run a local JSONL eval suite against semantic search.
 *
 * Reports recall@k and MRR metrics for semantic search quality evaluation.
 */
export function semanticEvalTools(ctx: PluginContext): Record<string, ToolDefinition> {
  const evalTool: ToolDefinition = {
    description: [
      "Run a local JSONL eval suite against semantic search and report recall@k and MRR.",
      "",
      "The JSONL file should have one case per line with 'query' and 'expected_paths' fields.",
      "",
      "Use when: evaluating semantic search quality, benchmarking retrieval, or after configuration changes.",
    ].join("\n"),
    args: {
      path: arg(z.string().describe("Path to the JSONL eval file (required)")),
      top_k: arg(
        z
          .number()
          .int()
          .positive()
          .optional()
          .describe("Maximum number of results per query (default: 10)"),
      ),
      include_per_case: arg(
        z.boolean().optional().describe("Include per-case results in response (default: true)"),
      ),
    },
    execute: async (args, _context): Promise<string> => {
      if (typeof args.path !== "string" || args.path.trim().length === 0) {
        throw new Error("semantic_eval: invalid params: `path` must be a non-empty string");
      }

      const params: Record<string, unknown> = {
        path: args.path,
      };

      if (typeof args.top_k === "number" && args.top_k > 0) {
        params.top_k = args.top_k;
      }

      if (typeof args.include_per_case === "boolean") {
        params.include_per_case = args.include_per_case;
      }

      const response = await callBridge(ctx, _context, "semantic_eval", params);

      if (response.success === false) {
        const message =
          typeof response.message === "string" && response.message.length > 0
            ? response.message
            : "semantic_eval failed";
        const code =
          typeof response.code === "string" && response.code.length > 0 ? response.code : undefined;
        throw new Error(code ? `semantic_eval: ${code} — ${message}` : message);
      }

      // Use Rust's summary_line if available
      if (typeof response.summary_line === "string" && response.summary_line.length > 0) {
        return response.summary_line;
      }

      // Build a readable summary from the structured response
      const parts: string[] = [];

      const total = typeof response.total === "number" ? response.total : 0;
      const hits = typeof response.hits_in_top_k === "number" ? response.hits_in_top_k : 0;
      const recall = typeof response.recall_at_k === "number" ? response.recall_at_k : 0;
      const mrr = typeof response.mrr === "number" ? response.mrr : 0;
      const k = typeof response.k === "number" ? response.k : 10;

      parts.push(`## Semantic Eval Results`);
      parts.push(`**Total cases:** ${total}`);
      parts.push(`**Hits in top-${k}:** ${hits}`);
      parts.push(`**Recall@${k}:** ${recall.toFixed(3)}`);
      parts.push(`**MRR:** ${mrr.toFixed(3)}`);

      // Per-case results
      const cases = response.cases;
      if (Array.isArray(cases) && cases.length > 0) {
        parts.push("\n### Per-Case Results");
        for (const c of cases) {
          const index = typeof c.index === "number" ? c.index : "?";
          const query = typeof c.query === "string" ? c.query : "unknown";
          const firstHitRank = typeof c.first_hit_rank === "number" ? c.first_hit_rank : null;
          const hit = typeof c.hit === "boolean" ? c.hit : false;

          let line = `- Case ${index}: "${query}"`;
          if (hit && firstHitRank !== null) {
            line += ` — hit at rank ${firstHitRank}`;
          } else {
            line += ` — no hit`;
          }
          parts.push(line);
        }
      }

      if (parts.length === 0) {
        return JSON.stringify(response, null, 2);
      }

      return parts.join("\n");
    },
  };

  return {
    aft_semantic_eval: evalTool,
  };
}
